File: 200_LeetCode_Questions/Java/D05P2_LeetCode_338.java
// Runtime: 1 ms, faster than 99.80% of Java online submissions for Counting Bits.
// Memory Usage: 43.5 MB, less than 73.28% of Java online submissions for Counting Bits.

class Solution {
    public int[] countBits(int num) {
        if(num == 0){
            return new int[]{0};
        }
        int ones[] = new int[num + 1];
        for(int i = 1; i < ones.length; i++){
            ones[i] = ones[i & (i - 1)] + 1;
        }
        return ones;
    }
}

File: 200_LeetCode_Questions/C++/D05P2_LeetCode_338.cpp
class Solution {
public:
    vector<int> countBits(int num) {
        vector<int> ans(num+1, 0);
        for(int i = 1; i <= num; i++) {
            if(i&1) {
                ans[i] = ans[i-1]+1;
            } else {
                ans[i] = ans[i/2];
            }
        }
        return ans;
    }
};
